src/main/java/com/google/sps/data/UserDatastoreRepository.java

```diff
@@ -0,0 +1,130 @@
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.sps.data.User;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+
+/**
+ * This class implements the match repository using the datastore 
+ */
+public class UserDatastoreRepository implements UserRepository {
+    
+  private Collection<User> allUsers = new ArrayList<>();
+  
+  private static final Gson gson = new Gson();
+
+  private DatastoreService datastore;
+  
+  private int maxProfiles = 10;
+  
+  public UserDatastoreRepository() {
+      datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  // function to add profile to database
+  public void addUserToDatabase(User user) {
+    String name = user.toString();
+    Collection<String> specialties = user.getSpecialties();
+
+    Entity userEntity = new Entity("User");
+    userEntity.setProperty("name", name);
+    if(specialties.size() > 0){
+        userEntity.setProperty("specialties", specialties);
+    }
+    datastore.put(userEntity);
+  }
+
+  // function to fetch the user from the database
+  public Collection<User> fetchUserProfiles() {
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+
+    Collection<User> userProfiles = new ArrayList<>();
+    for (Entity entity : resultsList) {
+        String name = (String) entity.getProperty("name");
+        Collection<String> specialties = (Collection) entity.getProperty("specialties");
+
+        User userObject = new User(name);
+        if(specialties != null) {
+            for (String specialty : specialties) {
+                userObject.addProductArea(specialty);
+            }
+        }
+        userProfiles.add(userObject);
+    }
+    return userProfiles;
+  }
+
+  public void removeUserProfile(User user){
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> resultsList = results.asList(FetchOptions.Builder.withLimit(maxProfiles));
+
+    String name = user.toString();
+
+    for (Entity entity : resultsList) {
+        String currentName = (String) entity.getProperty("name");
+        if(name.contains(currentName)) {
+            Key current = entity.getKey();
+            datastore.delete(current);
+        }
+    }
+  }
+
+  public void addUser(User user) {
+    allUsers = fetchUserProfiles();
```

Comment:
It looks like you're fetching all users from datastore each time you add a new user. Maybe fetch once on initialization?

Reply:
Im fetching it at top of function so I can check to see if the user exists, if I do it once before then that wouldn't update properly each time.

Reply:
Right, but fetching all users each time you add another is kind of expensive / won't scale well. Alternatively, you could just try and fetch the new user you're adding, and see if the response is null.

Related to your discussion about the allUsers variable above, the way I see it there's two ways of designing this class:
1- As a cache: You assume UserDatastoreRepository is the only object accessing your data in datastore anywhere in the world, and then load from datastore on construction, adding to it only when you have new users (essentially, you're treating UserDatastoreRepository as a cache)
2- As a wrapper: You don't assume you have full control over datastore, and thus fetch data as needed to insure it's up to date (e.g.: fetching a single user to see if they already exist).

Either works, but has different tradeoffs; Principally, 1 doesn't scale to multiple servers (though you'll likely not need this in this project) and you risk inconsistencies between what's stored and what's in memory and in datastore; 2 is what we use at google since everything we do requries multiple servers; It requires you to reason about all the ways the data may have changed in other servers, but since all data is always in datastore, there's no risk of your local copy being inconsistent (since there is no local copy).

Happy to discuss the tradeoffs more if you'd like. In any case, we probably want to avoid fetching all the users too frequently :)

Reply:
I see. I made a wrapper method for the datastore that it uses for adding and removing profiles.

Reply:
Done